Add clone phylogeny v1 (#152)
stjude-dnb-binfcore/sc-rna-seq-snap: analyses/clone-phylogeny-analysis/06A-create-numbat-plots.Rmd

# Consensus copy number segments

Let's take a look at the consensus segments.

```{r plot-2, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/",  sample_name[i], "/", "plot2.png")
p <- print(numbat_obj$plot_consensus())
ggsave(file = fname, p, width = 8, height = 6, device = "png")
```

# Bulk CNV profiles

We can also visualize these CNV events in pseudobulks where the data is more rich, aggregating cells by clone:

```{r plot-3, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot3.png")
p <- print(numbat_obj$bulk_clones %>% 

>   plot_bulks(min_LLR = 50, # filtering CNVs by evidence

Author: Fixed here: 359efa2ae2a069133fb81f270e8e2aac1e5d6e7f.

print(cnv_type)

umapCoord <- as.data.frame(Embeddings(object = seurat_obj@reductions[["umap"]])) # make dataframe with UMAP coordinates 
#numbat_obj.df <- as.data.frame(numbat_obj.df[[".__enclos_env__"]][["self"]][["allele_post"]])
#names(numbat_obj.df)[names(numbat_obj.df) == 'cell'] <- "UMI_id"
#head(numbat_obj.df)
metadata_table <- seurat_obj@meta.data
metadata_table$UMI_id <- rownames(metadata_table)
metadata_table <- merge.data.frame(metadata_table, numbat_obj.df, by = "UMI_id", all = TRUE)
```

## Joint posterior probability of the CNV (p_cnv)

`merge["p_cnv"]` contains cell-level information on specific CNV segments (seg), their alteration type (cnv_state), the joint posterior probability of the CNV (p_cnv), the expression-based posterior (p_cnv_x), and the allele-based posterior (p_cnv_y). We can visualize the event-specific posteriors in a expression-based embedding:

```{r plot-4, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
merge <- metadata_table
CellsMeta <- seurat_obj@meta.data
randomnumbers <- merge["p_cnv"]

CellsMeta["p_cnv"] <- randomnumbers
CellsMetaTrim <- subset(CellsMeta, select = c("p_cnv", cell_type_label))

> seurat_obj<- AddMetaData(seurat_obj, CellsMetaTrim)

Reviewer: Adding the `p_cnv` values to the `seurat_obj` metadata in this way does not work properly. The `numbat_obj$joint_post` data contains an entry for each CNV in each cell, so there are multiple entries for each cell in `metadata_table`. For example, for the sample `DYE2513` from the test data set, the `seurat_obj` has 4038 cells. The `metadata_table` has 64,608 entries (aka 4038 times 16 because there are 16 `CNV segs` reported in this sample). Essentially, this chunk of code is assigning the first 4038 `p_cnv` values in `randomnumbers` to the metadata in the `seurat_obj` and excluding the rest of the data. And even those first 4038 values assigned are not being assigned to the correct cells in the `seurat_obj` metadata, so the values being displayed on the UMAP below are incorrect for the cells.

I believe this is why the `Single-cell CNV calls` plotting in the `numbat` tutorial involves a step specifying which CNV `seg` to plot and creates a separate UMAP for each `seg`.

Reviewer: This same issue is applicable to the `p_cnv_x`and `p_cnv_y` metadata and plotting as well.

Author: Thank you for noticing this issue. There is a new solution for merging and plotting the data here: 4a4954e0816c8db7716d3db7c91b8e3d6367521e.

  ggtitle("Expression"))
ggsave(file = fname, p_cnv_x, width = 6, height = 5, device = "png")
```

## Allele-based posterior (p_cnv_y)

```{r plot-6, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
merge <- metadata_table
CellsMeta <- seurat_obj@meta.data
randomnumbers <- merge["p_cnv_y"]

CellsMeta["p_cnv_y"] <- randomnumbers
CellsMetaTrim <- subset(CellsMeta, select = c("p_cnv_y"))
seurat_obj <- AddMetaData(seurat_obj, CellsMetaTrim)

fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot6.png")
p_cnv_y <- print(FeaturePlot(seurat_obj, reduction = "umap", features = "p_cnv_y") +
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') + 
  ggtitle("Joint"))
ggsave(file = fname, p_cnv_y, width = 6, height = 5, device = "png") 
```

# CNVs and seg

```{r plot-7, fig.width = 20, fig.height = 15, fig.fullwidth = TRUE, echo=TRUE}
merge <- metadata_table
CellsMeta <- seurat_obj@meta.data
randomnumbers <- merge["seg"]

CellsMeta["seg"] <- randomnumbers
CellsMetaTrim <- subset(CellsMeta, select = c("seg"))

> seurat_obj <- AddMetaData(seurat_obj, CellsMetaTrim)

Reviewer: Similar to the issue with the metadata and plots above, this step seems to be adding a column to the metadata that repeats the list of the 16 CNV `seg` names until it reaches 4038. Plotting the UMAP below split by `seg` is essentially arbitrarily splitting the cells by which of the 16 segment labels was assigned in the metadata and coloring the cells by a `p_cnv` value which may or may not correspond to that actual cell and segment.

Author: Thank you for noticing this issue. There is a new solution for merging and plotting the data here: 4a4954e0816c8db7716d3db7c91b8e3d6367521e (as noted above).

fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot8.png")
p <- print(p1 + p2 + plot_layout(ncol = 2))
ggsave(file = fname, p, width = 15, height = 6, device = "png")
```

##  Genotypes vs Tumor microenvironment vs Cell type vs Disease stage

```{r plot-9, fig.width = 20, fig.height = 15, fig.fullwidth = TRUE, echo=TRUE}
library(RColorBrewer)
mypalette <- brewer.pal(12,"Paired") # colorblindFriendly = TRUE
mypalette <- colorRampPalette(mypalette)(40)

p3 <- DimPlot(seurat_obj, reduction = "umap", label = TRUE, group.by = cell_type_label, repel = TRUE, label.size = 2.5, cols = mypalette) + 
  ggtitle(cell_type_label)
p4 <- DimPlot(seurat_obj, reduction = "umap", group.by = 'Timepoint', label = TRUE, repel = TRUE, label.size = 2.5, cols = mypalette) + 
  ggtitle("Disease stage")

fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot9.png")
p <- print(p1 + p2 + p3 + p4 + plot_layout(ncol = 2, nrow=2))
ggsave(file = fname, p, width = 20, height = 15, device = "png")
```

# Tumor versus normal probability

```{r plot-10, fig.width = 20, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot10.png")

> p <- print(p_cnv + p_cnv_x + p_cnv_y + plot_layout(ncol = 3))

Reviewer: The `p_cnv`, `p_cnv_x`, and `p_cnv_y` values for the `numbat_obj$clone_post` and `numbat_obj$joint_post` are different values (with different meanings). It's confusing because the tool uses the same variable names across different data, but per the `numbat` [file descriptions](https://kharchenkolab.github.io/numbat/articles/descriptions.html), in `numbat_obj$clone_post` the `p_cnv...` values correspond to the probability of a cell belong to a clone with aneuploidy (aka being a malignant cell) and in `numbat_obj$joint_post` the values correspond to the probability of a cell having a specific CNV.  Here this code is plotting the values from `numbat_obj$joint_post`, but it should be plotting the values from  `numbat_obj$clone_post`.

Author: Thank you for noticing this issue. There is also addressed here: 4a4954e0816c8db7716d3db7c91b8e3d6367521e.

```{r plot-3, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot3.png")
p <- print(numbat_obj$bulk_clones %>% 
  plot_bulks(min_LLR = min_LLR_value, # filtering CNVs by evidence
             legend = FALSE))
ggsave(file = fname, p, width = 6, height = 5, device = "png")
```

# Single-cell CNV calls

Numbat probabilistically evaluates the presence/absence of CNVs in single cells. The cell-level CNV posteriors are stored in the `numbat_obj$joint_post` dataframe.

```{r echo=TRUE}
numbat_obj.df <- numbat_obj$joint_post %>% 
  select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y) %>%
  as.data.frame 

# Print cnv types
cnv_type <- numbat_obj.df %>% distinct(seg, cnv_state) %>% {setNames(.$cnv_state, .$seg)} 
print(cnv_type)

umapCoord <- as.data.frame(Embeddings(object = seurat_obj@reductions[["umap"]])) # make dataframe with UMAP coordinates 
metadata_table <- seurat_obj@meta.data
metadata_table$cell <- rownames(metadata_table)
metadata_table <- merge.data.frame(metadata_table, numbat_obj.df, by = "cell", all = TRUE) 
```

## Joint posterior probability of the CNV (p_cnv)

> `merge["p_cnv"]` contains cell-level information on specific CNV segments (seg), their alteration type (cnv_state), the joint posterior probability of the CNV (p_cnv), the expression-based posterior (p_cnv_x), and the allele-based posterior (p_cnv_y). We can visualize the event-specific posteriors in a expression-based embedding:

Reviewer: I think this edit makes the documentation slightly more clear and informative on where these values are coming from.

> `merge["p_cnv"]` contains cell-level information on specific CNV segments (seg), their alteration type (cnv_state), the joint posterior probability of the CNV (p_cnv), the expression-based posterior (p_cnv_x), and the allele-based posterior (p_cnv_y). We can visualize the event-specific posteriors in a expression-based embedding:

Author: Fixed here: fdbc6a04f16ca7fa8a241bf1bd3ac56c848fc5f3.

```{r p_cnv, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
# Extract UMAP coordinates
umap_data <- seurat_obj@reductions$umap@cell.embeddings
########################################################################
## p_cnv
# Extract feature data from metadata
feature_data_p_cnv <- metadata_table$p_cnv

# Combine UMAP coordinates and feature data into a data frame

> plot_data_p_cnv <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data_p_cnv)

Reviewer: These comments apply to the `p_cnv_x` and `p_cnv_y` metadata and plots as well.

> p_cnv <- ggplot(plot_data_p_cnv, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +

Reviewer: These comments apply to the `p_cnv_x` and `p_cnv_y` metadata and plots as well.

> p_cnv <- ggplot(plot_data_p_cnv, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +

Author: It is not an easy task to automate order by column name. This will remain as it is for now.

# Extract feature data from metadata
feature_data <- metadata_table$p_cnv_x

# Combine UMAP coordinates and feature data into a data frame
plot_data <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data)

p_cnv_x <- ggplot(plot_data, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  ggtitle("Expression") +
  theme_Publication () +
  theme(legend.position = "right")
########################################################################
## Allele-based posterior (p_cnv_y)
# Extract feature data from metadata
feature_data <- metadata_table$p_cnv_y

# Combine UMAP coordinates and feature data into a data frame
plot_data <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data)

p_cnv_y <- ggplot(plot_data, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +

>   ggtitle("Joint") +

Reviewer: ```suggestion
  ggtitle("Allele") +
```

Author: Fixed here: 22adedb63a3aa87854e1df135308a331490b4141.

  theme_Publication () +
  theme(legend.position = "right")
```  

```{r plot-4, fig.width = 20, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot4.png")

> p <- print(p_cnv + p_cnv_x + p_cnv_y + plot_layout(ncol = 3))

Reviewer: If you choose to facet these three plots by segment, then they should probably be plotted separately for readability.

Author: Won't fix for now.

ggsave(file = fname, p, width = 20, height = 6, device = "png")
```

# CNVs and seg

```{r plot-5, fig.width = 20, fig.height = 8, fig.fullwidth = TRUE, echo=TRUE}
split_data <- metadata_table$seg  # Replace 'seg' with your splitting variable

# Combine UMAP coordinates, feature data, and the split variable into a data frame
plot_data <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data_p_cnv, Seg = split_data)

# Create the plot using ggplot2
p1 <- ggplot(plot_data, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  facet_wrap(~Seg) +  # Split by 'seg' (or your splitting variable)
  theme_Publication () +

>   theme(legend.position = "right")

Reviewer: If you choose to facet the plots above by segment, then this plot could be removed as it would be a duplication of the `Joint` posterior probability plot above.

Author: I prefer to retain it for direct comparison purposes.

> p2 <- ggplot(plot_data_p_cnv, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +

Reviewer: This plot seems to be a duplicate of the current first plot in the `Joint posterior probability of the CNV` plots. And again, I think it is displaying too much overlapping data to be informative.

  ggtitle("Disease stage")

fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot7.png")
#p <- print(p1 + p2 + p3 + p4 + plot_layout(ncol = 2, nrow=2))
p <- print(p3 + p4 + plot_layout(ncol = 2, nrow=1))
ggsave(file = fname, p, width = 15, height = 6, device = "png")
```

# Tumor versus normal probability
```{r clone_post, echo=TRUE}
numbat_obj.df <- numbat_obj$clone_post 

# Extract UMAP coordinates
umap_data <- seurat_obj@reductions$umap@cell.embeddings

##############################################################
# Extract feature data from metadata
feature_data_p_cnv <- numbat_obj.df$p_cnv
# Combine UMAP coordinates and feature data into a data frame
plot_data_p_cnv <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data_p_cnv)

# Create the plot using ggplot2
p_cnv <- ggplot(plot_data_p_cnv, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +

>   ggtitle("Allele") +

Reviewer: ```suggestion
  ggtitle("Joint") +
```

Author: Can you cite the basis of this suggestion? BTW, it is addressed here: fdbc6a04f16ca7fa8a241bf1bd3ac56c848fc5f3.

##############################################################
# Extract feature data from metadata
feature_data <- numbat_obj.df$p_cnv_x

# Combine UMAP coordinates and feature data into a data frame
plot_data <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data)

p_cnv_x <- ggplot(plot_data, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  ggtitle("Expression") +
  theme_Publication () +
  theme(legend.position = "right")
##############################################################
# Extract feature data from metadata
feature_data <- numbat_obj.df$p_cnv_y

# Combine UMAP coordinates and feature data into a data frame
plot_data <- data.frame(UMAP_1 = umap_data[, 1], UMAP_2 = umap_data[, 2], Feature = feature_data)

p_cnv_y <- ggplot(plot_data, aes(x = UMAP_1, y = UMAP_2, color = Feature)) +
  geom_point(alpha = 0.8) +  # Adjust transparency if needed
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +

>   ggtitle("Joint") +

Reviewer: ```suggestion
  ggtitle("Allele") +
```

Author: Can you cite the basis of this suggestion? BTW, it is addressed here: fdbc6a04f16ca7fa8a241bf1bd3ac56c848fc5f3.

# Define and order palette
mypalette <- mypalette_df$hex_codes 
names(mypalette) <- mypalette_df$timepoint
```

```{r define-parameters-for-plots, echo=TRUE}
# Read color palette
palette_df <- readr::read_tsv(cell_types_palette_file, guess_max = 100000, show_col_types = FALSE) %>%
  mutate(cell_type_names = case_when(cell_type_names == "na_color" ~ "unassigned",
                                       TRUE ~ cell_type_names))

# Define and order palette
palette <- palette_df$hex_codes 
names(palette) <- palette_df$cell_type_names 
```

# Copy number landscape and single-cell phylogeny

As an overview, we can visualize the CNV calls in single-cells and their evolutionary relationships in an integrated plot panel:

```{r plot-1, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot1.png")
p <- print(numbat_obj$plot_phylo_heatmap(clone_bar = TRUE, 
                                         p_min = 0.9,
                                         pal_clone = mypal))
ggsave(file = fname, p, width = 8, height = 6, device = "png")
```

> In this visualization, the single-cell phylogeny (left) is juxtaposed with a heatmap of single-cell CNV calls (right). The CNV calls are colored by the type of alterations (AMP, amplification, BAMP, balanced amplification, DEL, deletion, CNLoH, copy-neutral loss of heterozygosity). The colorbar in-between differentiates the distinct genetic populations (genotype). The dashed blue line separates the predicted tumor versus normal cells. This tells us that the dataset mainly consists of three cell populations, a normal population (gray) and two tumor subclones (green and purple).

Author: See 883388e18fa0bf782e3f54234be368426cba1e10.

  ggtitle("Joint posterior probability of CNV state given expression data") +
  theme_Publication () #+
  #theme(legend.position = "right")

########################################################################
## Allele-based posterior (p_cnv_y)
p_cnv_y <- ggplot(metadata_table, aes(x = UMAP_1, y = UMAP_2, color = p_cnv_y)) +
  geom_point(alpha = 1, size = 1) +  # Adjust transparency if needed: alpha = 0.8; # Setting alpha to 1 for no transparency
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  ggtitle("Joint posterior probability of CNV state given allele data") +
  theme_Publication () +
  theme(legend.position = "right")
```  

```{r plot-4, fig.width = 25, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot4.png")
p <- print(p_cnv + p_cnv_x + p_cnv_y + plot_layout(ncol = 3))
ggsave(file = fname, p, width = 25, height = 6, device = "png")
```

# CNVs and seg

```{r plot-5, fig.width = 20, fig.height = 8, fig.fullwidth = TRUE, echo=TRUE}
# Sort in descending order, you can use desc()

> plot_data <- metadata_table %>%
>   arrange(seg)

Reviewer: ```suggestion
plot_data <- metadata_table
plot_data$seg <- factor(plot_data$seg, levels = names(cnv_type))
```

I tested and this code works for ordering the plots faceted by segment, so maybe consider an update like this for the next version of the module.

Author: Thanks! Fixed here: 0ce34fab327c0a234b94ba67082bf341aee145cd. Please notice this version is more idiomatic and concise in the tidyverse style, making the code more readable .

  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  facet_wrap(~seg) +  # Split by 'seg' (or your splitting variable)
  theme_Publication () +
  theme(legend.position = "right")

p2 <- ggplot(metadata_table, aes(x = UMAP_1, y = UMAP_2, color = p_cnv)) +
  geom_point(alpha = 1, size = 1) +  # Adjust transparency if needed: alpha = 0.8; # Setting alpha to 1 for no transparency
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior') +
  ggtitle("Joint posterior probability of CNV state: All identified CNVs") +
  theme_Publication () +
  theme(legend.position = "right")

fname <- paste0(numbat_plots_dir, "/", sample_name[i], "/", "plot5.png")
p <- print(p1 + p2 + plot_layout(ncol = 2, nrow = 1))
ggsave(file = fname, p, width = 20, height = 8, device = "png")
```

# Clonal assignments

Numbat aggregates signals across subclone-specific CNVs to probabilistically assign cells to subclones. The information regarding clonal assignments are contained in the `numbat_obj.df$clone_post` dataframe.

> Here clone_opt denotes the maximum likelihood assignment of a cell to a specific clone. p_{1..4} are the detailed breakdown of the posterior probability that the cell belongs to each clone, respectively. Let's visualize the clonal decomposition in a tSNE embedding. Note that clone 1 is always the normal cells.

Reviewer: I think this section could be edited to be a bit less specific and more applicable to datasets with more than four clones.

```suggestion
Here clone_opt denotes the maximum likelihood assignment of a cell to a specific clone. p_{1..n} are the detailed breakdown of the posterior probability that the cell belongs to each clone, i.e. clone 1 to clone n, respectively. Let's visualize the clonal decomposition in a tSNE embedding. Note that clone 1 is always the normal cells.
```

Author: See 0ce34fab327c0a234b94ba67082bf341aee145cd.
